File: backend/src/main/java/edu/cit/caleza/washtrack/features/profile/ProfileService.java
package edu.cit.caleza.washtrack.features.profile;

import edu.cit.caleza.washtrack.shared.entity.User;
import edu.cit.caleza.washtrack.shared.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class ProfileService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    // FR-003: fetch current profile
    public ProfileResponse getProfile(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));
        return toResponse(user);
    }

    // FR-003: update name, phone, address (email is read-only)
    public ProfileResponse updateProfile(Long userId, ProfileRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));

        user.setFullName(request.getFullName());
        user.setPhone(request.getPhone());
        user.setAddress(request.getAddress());

        User saved = userRepository.save(user);
        return toResponse(saved);
    }

    // FR-003: change password — requires the correct current password first
    public void changePassword(Long userId, ProfilePasswordRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));

        if (!passwordEncoder.matches(request.getCurrentPassword(), user.getPasswordHash())) {
            throw new IllegalArgumentException("Current password is incorrect");
        }

        user.setPasswordHash(passwordEncoder.encode(request.getNewPassword()));
        userRepository.save(user);
    }

    private ProfileResponse toResponse(User user) {
        return new ProfileResponse(
                user.getUserId(),
                user.getFullName(),
                user.getEmail(),
                user.getPhone(),
                user.getAddress(),
                user.getRole().name(),
                user.getCreatedAt()
        );
    }
}

File: web/src/features/profile/profileService.js
import axiosClient from "../../api/axiosClient";

// FR-003: fetch the full profile for the given user
export async function getProfile(userId) {
  const response = await axiosClient.get(`/profile/${userId}`);
  return response.data;
}

// FR-003: update name, phone, address (email is read-only)
export async function updateProfile(userId, profileData) {
  const response = await axiosClient.put(`/profile/${userId}`, profileData);
  return response.data;
}

// FR-003: change password — requires the current password for verification
export async function changePassword(userId, passwordData) {
  const response = await axiosClient.put(`/profile/${userId}/change-password`, passwordData);
  return response.data;
}
